research/delf/delf/python/datasets/tuples_dataset.py

```diff
@@ -0,0 +1,356 @@
+      imsize: Integer, defines the maximum size of longer image side transform.
+      nnum: Integer, number of negative images for a query image in a
+        training tuple.
+      qsize: Integer, number of query images to be processed in one epoch.
+      poolsize: Integer, size of the negative image pool, from where the
+        hard-negative images are re-mined.
+      loader: Callable, a function to load an image given its path.
+
+    Raises:
+      ValueError: If mode is not either 'train' or 'val'.
+    """
+
+    if mode not in ['train', 'val']:
+      raise ValueError(
+              "`mode` argument should be either 'train' or 'val', passed as a "
+              "String.")
+
+    # Loading db.
+    db_fn = os.path.join(data_root, '{}.pkl'.format(name))
+    with tf.io.gfile.GFile(db_fn, 'rb') as f:
+      db = pickle.load(f)[mode]
+
+    # Initializing tuples dataset.
+    self._ims_root = data_root if not ims_root else ims_root
+    self._name = name
+    self._mode = mode
+    self._imsize = imsize
+    self._clusters = db['cluster']
+    self._qpool = db['qidxs']
+    self._ppool = db['pidxs']
+
+    if not hasattr(self, 'images'):
+      self.images = db['ids']
+
+    # Size of training subset for an epoch.
+    self._nnum = nnum
+    self._qsize = min(qsize, len(self._qpool))
+    self._poolsize = min(poolsize, len(self.images))
+    self._qidxs = None
+    self._pidxs = None
+    self._nidxs = None
+
+    self._loader = loader
+    self._print_freq = 10
+    self._n = 0
+
+  def __iter__(self):
+    """Function for making TupleDataset an iterator.
+
+    Returns:
+      iter: The iterator object itself (TupleDataset).
+    """
+    return self
+
+  def __next__(self):
+    """Function for making TupleDataset an iterator.
+
+    Returns:
+      next: The next item in the sequence (next dataset image tuple).
+    """
+    if self._n < len(self._qidxs):
+      result = self.__getitem__(self._n)
+      self._n += 1
+      return result
+    else:
+      raise StopIteration
+
+  def _img_names_to_full_path(self, image_list):
+    """Converts list of image names to the list of full paths to the images.
+
+    Args:
+      image_list: List of image names.
+
+    Returns:
+      image_full_paths: List of full paths to the images.
+    """
+    if not isinstance(image_list, list):
+      return os.path.join(self._ims_root, image_list)
+    return [os.path.join(self._ims_root, img_name) for img_name in image_list]
+
+  def __getitem__(self, index):
+    """Called to load an image at the given `index`.
+
+    Args:
+      index: Integer, index.
+
+    Returns:
+      output: Tuple [q,p,n1,...,nN, target], loaded 'train'/'val' tuple at
+        index of qidxs. `q` is the query image tensor, `p` is the
+        corresponding positive image tensor, `n1`,...,`nN` are the negatives
+        associated with the query. `target` is the tensor of labels
+        corresponding to the tuple list: query (-1), positive (1), negative (0).
```

Comment:
just to confirm my understanding here, `target` is an integer 1D tensor of shape [2+N] -- is this correct? If so, maybe we could mention these additional details in the docstring (*integer*, its shape)